### PR TITLE
3135 - 3 WIP

### PR DESCRIPTION
**Pull request by nikolaiprivalov**
## Номер
3135

## Номер задания
3

## Комментарии

### CODE REVIEW

File: 3135/3/helpers/utils.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    end
+  end
+
+  def registration(params)
+    user = User.new(params)
+    if user.save
+      activate_user(user.id)
+      session[:return_to_page]
+    else
+      flash.next[:error] = '!!wrong input!!'
```
Comment (iris-the-fox) [inline 3135/3/helpers/utils.rb]: more informative messages

File: 3135/3/views/restaurant_in_detail.erb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<div>
+    <hr>
+    <h4><%= @restaurant.name %></h4>
+    <p><%= @restaurant.location %></p>
+    <p><%= @restaurant.description %></p>
+    <h4>Average rating: <%= @average_rating.to_s %></h4>
```
Comment (iris-the-fox) [inline 3135/3/views/restaurant_in_detail.erb]: Expression tags converts to a string before rendering, noneed to use to_s method

File: 3135/3/views/layout.erb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<html>
+    <head>
+        <title>MUNCH BUNCH</title>
+    </head>
+    <body>
+    
+    <div>
+        <% if @active_user %>
+                <%= @active_user.username %>
```
Comment (IvanShamatov) [inline 3135/3/views/layout.erb]: html indentation is inconsistent. How about using 2 spaces in HTML too?

File: 3135/3/app.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+require 'sinatra/activerecord'
+require 'sinatra/flash'
+require 'pry'
+
+set :database, 'sqlite3:db/database.sqlite3'
+
+enable :sessions
+
+before do
+  @active_user = User.find_by(id: session[:id])
```
Comment (IvanShamatov) [inline 3135/3/app.rb]: https://stackoverflow.com/a/43901290/703725

File: 3135/3/app.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+  pass if ['/login', '/logout', '/registration', '/add_review'].include? request.path_info
```
Comment (IvanShamatov) [inline 3135/3/app.rb]: That's a bad idea. So there will be a moment when you forget about adding some new URL to this filter

Comment (IvanShamatov) [reply 3135/3/app.rb]: https://github.com/gnab/sinatra-authorize here are some thoughts on authorization

File: 3135/3/app.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+post '/add_review' do
+  review = Review.new(params)
+  flash[:error] = review.errors.full_messages.join('; ') unless review.save
+
+  redirect session[:return_to_page]
+end
+
+# restaurants_in_detail
+get '/:name' do
+  @restaurant = Restaurant.find_by(name: params[:name])
```
Comment (IvanShamatov) [inline 3135/3/app.rb]: unfortunately, making this name unique without the ability to add spaces, slashes, special characters in the name, looks like a bad idea.
If you would like to add readable slug, why not trying gem `friendly_id` or similar
I also recommend you to use `get '/restaurants/:name'` for namespacing reasons

File: 3135/3/views/restaurant_in_detail.erb [hunk 7]
```diff
@@ -0,0 +1,45 @@
+<div>
+    <hr>
+    <h4><%= @restaurant.name %></h4>
+    <p><%= @restaurant.location %></p>
+    <p><%= @restaurant.description %></p>
+
+    <% if @average_rating %>
+        <h4>Average rating: <%= @average_rating.round(1) %></h4>
```
Comment (drujok) [inline 3135/3/views/restaurant_in_detail.erb]: в контроллер @average_rating.round(1)

File: 3135/3/app.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  if user
+    session[:id] = user.id
+    redirect session[:return_to_page]
+  else
+    flash[:error] = 'Invalid email/password combo!'
+    redirect '/login'
+  end
+end
+
+get '/logout' do
```
Comment (drujok) [inline 3135/3/app.rb]: post

File: 3135/3/views/layout.erb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+      <% flash.keys.each do |type| %>
+        <% if type == :notice %>
+        <%= flash[:notice] %>
+        <% end %>
+
+        <% if type == :error %>
+        <%= flash[:error].full_messages.join('; ') %>
+        <% end %>
+      <% end %>
```
Comment (IvanShamatov) [inline 3135/3/views/layout.erb]:
```ruby
<%= flash[:notice] if flash.key?(:notice) %>
<%= flash[:error].full_messages.join('; ') if flash.key?(:error) %>
```

**Pull request merged**